File: sdk/python/src/axcp/replay.py
"""Thread-safe sequence and nonce replay protection."""

from __future__ import annotations

import dataclasses
import threading
from collections import defaultdict
from datetime import datetime, timedelta, timezone

from axcp.errors import InvalidReplayConfigError, ReplayDetectedError, SequenceTooOldError


@dataclasses.dataclass
class _PeerWindow:
    highest: int = 0
    seen: dict[int, datetime] = dataclasses.field(default_factory=dict)


class ReplayProtector:
    """Sequence and nonce replay guard with TTL cleanup."""

    def __init__(self, ttl: timedelta = timedelta(minutes=5), window: int = 64) -> None:
        if ttl <= timedelta(0):
            raise InvalidReplayConfigError("ttl must be positive")
        if window < 1:
            raise InvalidReplayConfigError("window must be at least 1")
        self._ttl = ttl
        self._window = window
        self._peers: dict[str, _PeerWindow] = {}
        self._nonces: dict[str, dict[str, datetime]] = defaultdict(dict)
        self._lock = threading.Lock()

    def check_and_mark(self, peer_id: str, sequence: int, now: datetime | None = None) -> None:
        if not peer_id:
            raise InvalidReplayConfigError("peer_id cannot be empty")
        if sequence < 0:
            raise InvalidReplayConfigError("sequence cannot be negative")

        current = now or datetime.now(tz=timezone.utc)
        with self._lock:
            window = self._peers.setdefault(peer_id, _PeerWindow())
            self._cleanup_sequences(window, current)

            if sequence in window.seen:
                raise ReplayDetectedError("sequence already seen")

            if window.highest > 0:
                min_acceptable = (
                    window.highest - self._window + 1
                    if window.highest >= self._window
                    else 0
                )
                if sequence < min_acceptable:
                    raise SequenceTooOldError("sequence outside replay window")

            window.seen[sequence] = current
            if sequence > window.highest:
                window.highest = sequence

    def check_and_mark_nonce(self, peer_id: str, nonce: str, now: datetime | None = None) -> None:
        if not peer_id:
            raise InvalidReplayConfigError("peer_id cannot be empty")
        if not nonce:
            raise InvalidReplayConfigError("nonce cannot be empty")

        current = now or datetime.now(tz=timezone.utc)
        with self._lock:
            peer_nonces = self._nonces[peer_id]
            self._cleanup_nonces(peer_nonces, current)
            if nonce in peer_nonces:
                raise ReplayDetectedError("nonce already seen")
            peer_nonces[nonce] = current

    def _cleanup_sequences(self, window: _PeerWindow, now: datetime) -> None:
        cutoff = now - self._ttl
        expired = [sequence for sequence, seen_at in window.seen.items() if seen_at < cutoff]
        for sequence in expired:
            del window.seen[sequence]

    def _cleanup_nonces(self, nonces: dict[str, datetime], now: datetime) -> None:
        cutoff = now - self._ttl
        expired = [nonce for nonce, seen_at in nonces.items() if seen_at < cutoff]
        for nonce in expired:
            del nonces[nonce]

File: sdk/typescript/src/replay.ts
import {
  InvalidReplayConfigError,
  ReplayDetectedError,
  SequenceTooOldError,
} from "./errors.js";

interface PeerWindow {
  highest: number;
  seen: Map<number, number>;
}

export interface ReplayProtectorOptions {
  readonly ttlMs?: number;
  readonly window?: number;
}

export class ReplayProtector {
  private readonly ttlMs: number;
  private readonly window: number;
  private readonly peers = new Map<string, PeerWindow>();
  private readonly nonces = new Map<string, Map<string, number>>();

  constructor(options: ReplayProtectorOptions = {}) {
    this.ttlMs = options.ttlMs ?? 5 * 60 * 1000;
    this.window = options.window ?? 64;
    if (this.ttlMs <= 0) {
      throw new InvalidReplayConfigError("ttlMs must be positive");
    }
    if (this.window < 1) {
      throw new InvalidReplayConfigError("window must be at least 1");
    }
  }

  checkAndMark(peerId: string, sequence: number, now: Date = new Date()): void {
    if (!peerId) {
      throw new InvalidReplayConfigError("peerId cannot be empty");
    }
    if (!Number.isSafeInteger(sequence) || sequence < 0) {
      throw new InvalidReplayConfigError("sequence must be a non-negative safe integer");
    }

    const currentMs = now.getTime();
    const window = this.getPeerWindow(peerId);
    this.cleanupSequences(window, currentMs);

    if (window.seen.has(sequence)) {
      throw new ReplayDetectedError("sequence already seen");
    }

    if (window.highest > 0) {
      const minAcceptable = window.highest >= this.window ? window.highest - this.window + 1 : 0;
      if (sequence < minAcceptable) {
        throw new SequenceTooOldError("sequence outside replay window");
      }
    }

    window.seen.set(sequence, currentMs);
    if (sequence > window.highest) {
      window.highest = sequence;
    }
  }

  checkAndMarkNonce(peerId: string, nonce: string, now: Date = new Date()): void {
    if (!peerId) {
      throw new InvalidReplayConfigError("peerId cannot be empty");
    }
    if (!nonce) {
      throw new InvalidReplayConfigError("nonce cannot be empty");
    }

    const currentMs = now.getTime();
    const peerNonces = this.getPeerNonces(peerId);
    this.cleanupNonces(peerNonces, currentMs);
    if (peerNonces.has(nonce)) {
      throw new ReplayDetectedError("nonce already seen");
    }
    peerNonces.set(nonce, currentMs);
  }

  private getPeerWindow(peerId: string): PeerWindow {
    const existing = this.peers.get(peerId);
    if (existing !== undefined) {
      return existing;
    }
    const created = { highest: 0, seen: new Map<number, number>() };
    this.peers.set(peerId, created);
    return created;
  }

  private getPeerNonces(peerId: string): Map<string, number> {
    const existing = this.nonces.get(peerId);
    if (existing !== undefined) {
      return existing;
    }
    const created = new Map<string, number>();
    this.nonces.set(peerId, created);
    return created;
  }

  private cleanupSequences(window: PeerWindow, nowMs: number): void {
    const cutoff = nowMs - this.ttlMs;
    for (const [sequence, seenAt] of window.seen) {
      if (seenAt < cutoff) {
        window.seen.delete(sequence);
      }
    }
  }

  private cleanupNonces(nonces: Map<string, number>, nowMs: number): void {
    const cutoff = nowMs - this.ttlMs;
    for (const [nonce, seenAt] of nonces) {
      if (seenAt < cutoff) {
        nonces.delete(nonce);
      }
    }
  }
}
